### PR TITLE
WIP:Version3 roadmap and taklist

### PR DESCRIPTION
**Pull request by vrajroham**
---

## Tasklist / Issues
- [x] vue-dropzone without a form wrapper. #196 
- [x] Overwritten function `accept` doesn't get called. #186 
- [x] Get rid of `props` as options, just use `options object`.
- [x] Request - Option to disable injected <style>. #151 
- [x] `manuallyAddFile` causes one additional redundant http image call. #192 
- [x] Not all DropzoneJS configuration is supported. #125 
- [x] Nest dropzone in other forms. #114 
- [x] Send additional parameters with images. 
- [x] Custom template is using component styles instead of overridden styles. #168 
- [x] Multiple icon issue (font-awesome, material)
- [x] README file contents.
- [x] Edit link on each docs page. It will be easy for users to improve docs and submit small patches, because from link the only file will be served to user will which needs to-be modified.
- [x] Tutorial to demonstrate AWS S3 feature. 

## Needs more info
- [x] Upload files directly to AWS S3.  (Need some more info as to what is required...)
- [x] Upload files using REST API #161 (I dont really understand this?)
- [x] Override component css (is this addressed above by #151)

---
> @rowanwins edit the issue according to your suggestions

### CODE REVIEW

**Comment by rowanwins** [comment]
This is looks great thanks @vrajroham ! I'll go through the list tonight and tick some of them off because Im sure a number of them are addressed.

**Comment by rowanwins** [comment]
Hey @vrajroham 

I reckon this is nearly ready to go. 

There are a couple of items I'm not sure about in the list above, I'm not sure if you understand them better than me perhaps otherwise we can leave them I think? 

I'm also waiting to see if #204 can be incorporated, I'm not sure if you have much experience with slots but I couldn't make this work which was a shame...

The docs could probably be tidied a little bit, particularly some of the demo pages, but I think they are looking pretty good.

I also want to do a quick test of the built version to check that I can use it (given that the demo page uses the component from the `src` directory.

If anything else comes to mind let me know. 

And also thanks to @AlexanderYW for helping kick things along, it's always great when multiple people contribute to help spread the load 👍

**Comment by vrajroham** [comment]
Hi @rowanwins 

Very firstly I would like to say please don't say below things. You are awesome buddy....
>  I'm not sure if you have much experience with slots but I couldn't make this work which was a shame..

There's nothing shameful of not knowing, if so then I should be more shameful because I too don't know much things which I learned from you. Yet there's lot learn from you and work with you. :wink: 

I too don't have experience with `slots` but I'll give try to it today, will see if I can do it.

**Yes!** :+1:  We really appreciate your contribution @AlexanderYW . Hoping to work more with you if possible.

----
### About task list

- I can confirm that  **Upload files using REST API #161 ** is not an issue or feature or modification. It's just matter of `POST` or `PUT` method that's it. No need to work on that. **I marked it as done**
- **Direct S3 upload** is just my idea which we can consider as feature.

### Questions

- How are we going to host docs?
- What will be status of V2.* after release of **V3**?

---
I think you give a try on Monday or later. I didn't got much time to go through component after your modifications. In a day to two I'll look on to it and do some tweaks if necessary, then we can test it. 

What are your thoughts?

**Comment by vrajroham** [comment]
Hey @rowanwins 
I spend hours to make slots work, but seems can't be done yet. Because we are passing previewTemplate to Dropzone as a function that returns a string, so you probably can't do that when there are slots involved (Vue doesn't manage the internals of Dropzone)

We need to drop this here I think, because already there's way to achieve this without `slots` using `previewTemplate` option. Let it remain as it is.

**Comment by vrajroham** [comment]
@rowanwins It really sucks to modify default styles of dropzone. 
I created custom preview template and tried to change styles. Most of the styles are not changing and weird behaviour of success, error icons etc. I tried `less` then `css` but no luck. It compulsory uses all styles of core dropzone library.
Looking for other alternative.

**Comment by rowanwins** [comment]
Hi @vrajroham 

Thanks for taking a look into the slots thing, I agree with your assessment to leave it as is at the moment.

I'll look into the styling issue tonight and see what I can work out.

**Comment by AlexanderYW** [comment]
@vrajroham if you help me understand what you want to achieve with the modifying of the default styles maybe I can help :)

**Comment by rowanwins** [comment]
Gday @AlexanderYW 

So basically if someone wants to completely change the look of the component we want them to be able to do so. 
Part 1 is that they should be able to pass a custom preview template using the [previewTemplate property](http://www.dropzonejs.com/#layout). 
Part 2 is that they should be able to use whatever css they want to make things look how they want. 

If you're happy to take a look that would be great otherwise I can take a look in a few days.

**Comment by vrajroham** [comment]
@AlexanderYW The same thing just @rowanwins said above. I got little success in [this](https://github.com/rowanwins/vue-dropzone/blob/version3/docs/src/pages/SlotsDemo.vue) component and you can try there.

**Comment by vrajroham** [comment]
Soo Sorry! My bad, I mistakenly deleted v3-master branch.

**Comment by vrajroham** [comment]
If @rowanwins  you have `v3-master` locally pulled then please push it again.

**Comment by vrajroham** [comment]
@rowanwins I restored the branch! 

Thanks

**Comment by rowanwins** [comment]
Hey @vrajroham 

Don't worry about :) I can't count the amount of times I've stuffed things up on git! 

I think you should be able to grab the files from this page
https://github.com/rowanwins/vue-dropzone/pull/205/commits
Just scroll to the last commit and I think you should be able to download the files.

**Comment by vrajroham** [comment]
Hey @rowanwins 
> I also want to do a quick test of the built version to check that I can use it (given that the demo page uses the component from the src directory.

You can proceed for test..

**Comment by rowanwins** [comment]
HI @vrajroham 

I've just taken a bit of a look at the s3 stuff you've added. My only real hesitation is adding another dependency by adding axios (which add 12kb). If you could perhaps refactor it to remove that dependency and just do a standard[ xhr post request ](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/send)that would be awesome...

Great job on the custom preview template though!

**Comment by vrajroham** [comment]
@rowanwins I'll update it!

Thanks :smile:

**Comment by r4fx** [comment]
> I can confirm that Upload files using REST API #161 is not an issue or feature or modification. It's just matter of POST or PUT method that's it. No need to work on that. I marked it as done

@vrajroham could you provide an example how to achieve this ? I will be appreciate.

**Comment by vrajroham** [comment]
Sure @r4fx why not! Can you provide the scenario about what you want to achieve? Or I would share how I achieved? So I can explain better.

**Comment by r4fx** [comment]
In my scenario i have an app with authorization, user needs to login to have access to application tools, my problem with vue-dropzone is that i can't (or i don't know how) to get response status during upload a file, i need to get for example 401 status to logout user if user session ended.
All my actions in app are using Axios (with Authorization inside interceptors), and there i can catch error with proper status, but not in case with vue-dropzpne where `uploadUrl` is a simple string url.

One option which do the job is add a hook to AJAX request (https://stackoverflow.com/a/5202999/1816520) but i'm looking for cleaner solution.

**Comment by rowanwins** [comment]
Hi @vrajroham & @AlexanderYW 

I've overhauled the custom preview demo so if you'd like to take a look that would be great.
FInal thing I want to test to pulling in the built file into a project and checking it all works ok.
Cheers

**Comment by vrajroham** [comment]
Hey guys eager to have a look on library but currently I'm  at hometown
 for festival till Sunday. Hope you guys are well!

**Comment by vrajroham** [comment]
@rowanwins If possible have a look on below github button in sidebar. This button is not clickable in Ubuntu.

screen shot 2017-10-24 at 11 14 02 pm

**Comment by rowanwins** [comment]
Alright @vrajroham & @AlexanderYW I reckon this is all done and ready for release.

- This evening I've tested the build process and that all seems to work nicely when importing it into another project, I've added a couple more instructions to the installation page
- The docs get built using `npm run build-docs` and end up in `/docs/dist`, once we merge this branch I'll update this github repo link to the new docs

So final call for review/comments before I merge it! 

Thanks so much for your help in getting it this far. I'm pretty excited about how it's turned out :)

File: docs/src/pages/demo.vue [hunk 1]
```diff
@@ -156,13 +158,23 @@
 </template>
 
 <script>
-import { vueDropzone } from '../../../src/';
+import vueDropzone from '../../../src/';
 import editDoc from '../components/DocEditLink.vue';
+var example = `
+<vue-dropzone ref="myVueDropzone" id="dropzone" :options="dropzoneOptions">
+....
+dropzoneOptions: () {
```
Comment (vrajroham) [inline docs/src/pages/demo.vue]: Is it correct or typo?

**Comment by AlexanderYW** [comment]
I do a complete review later today around 20.00 danish time (GMT +1)

**Comment by vrajroham** [comment]
@rowanwins Sounds great as everything is working as expected! Kudos to team :1st_place_medal: .

**Comment by AlexanderYW** [comment]
I dont really have any comments beside (let me know if you disagree), that we should split up the example as I did on `customPreviewDemo.vue` and that I forgot to close the `vue-dropzone` tag.

If you agree, would any of you mind correcting it?, else I can do it on sunday :)

**Comment by vrajroham** [comment]
@AlexanderYW I fixed the typo in 15f1086 . It's good idea to split the examples, it will surely help newbies to understand and our docs will be very easy to understand and maintain! Good thought, we can make every example page same! I'll update it tonight.

Thanks :smile:

**Comment by Fohlen** [comment]
I would like to add 

- [ ] template slots to the list

I will fork of this branch and go on with my PR from there.

**Comment by vrajroham** [comment]
Hi @Fohlen 
We tried some alternative ways to achieve it. It will be awesome if you can accomplish it!

**Comment by Fohlen** [comment]
hey @vrajroham 
I didn't test this entirely, but it's a very well proof of concept 👍 
https://github.com/Fohlen/vue-dropzone/commit/493260eee0e2e5c765e3e0a98f642e7ce4d772a0

**Comment by vrajroham** [comment]
@Fohlen I'll take a look on your work and revert back here.

**Comment by vrajroham** [comment]
🎉 Hurray.... 
Finally merged. Hey @rowanwins I would suggest make **master** Branch as default instead of **v3-master**. What you think?

**Comment by rowanwins** [comment]
Its alive @vrajroham , yep I've done as suggested :) 

Now published to npm as well!

**Comment by vrajroham** [comment]
@rowanwins Tonight I'm going to do following updates, I want your confirmation on following.
* Add/list our component to base/core dropzone site. Already multiple projects are listed as shown in attached image.
* Opening PR for sub domain on [js.org](js.org). Suggest the name for our library.
screen shot 2017-10-29 at 3 43 45 pm

**Comment by rowanwins** [comment]
Hi @vrajroham 

Yep feel free to do either or both of those things. I'm not sure what we'd use the sub-domain for but if you feel like it serves a purpose go for it (I'd suggest vue2-dropzone) :)

**Pull request merged**